### ISSUE TITLE
GIT_SYNC_ROOT is persisted somewhere on disk it seems

### ISSUE THREAD

**Issue by yuranos** [kubernetes/git-sync#334]
I haven't thoroughly checked the codebase for now, but I will describe what's happening in my case.

I'm using git-sync as initContainer to preprovision a disk based on github content.
I'm mounting a persistent volume to this initContainer and point `GIT_SYNC_ROOT` to it.
In my case, it was initially `/gitrepo`.

But later on I decided to change the mount pass so that it's the same between the main container and the init one.

```
- name: GIT_SYNC_ROOT
              value: /gitrepo
```

to:

```
- name: GIT_SYNC_ROOT
              value: /tmp
```

And: 
`mountPath: /gitrepo`
to:
`mountPath: /tmp`

So, basically I don't have /gitrepo anywhere in my k8s declarations any longer.

Imagine my surprise when I get:
`E0208 22:29:09.673940      10 main.go:348]  "msg"="failed to sync repo, aborting" "error"="error running command: exit status 128: \"fatal: Invalid path '/gitrepo': No such file or directory\\nPreparing worktree (detached HEAD cc98f6d5)\\n\"" `

After I reprovisioned the disk, everything started to work again.

@thockin , please don't tell me to use git clone instead. I really like this tool:)

**Comment by thockin** [comment]
AFAIK it is not persisted anywhere - we don't have a checkpoint or anything
and the only place we ever look for it on disk is specified by the
flags/env.

Can you repro it?  Show me a minimized case and I can at least shell in and
see what happens with verbose logs.

**Comment by yuranos** [comment]
I'm trying a minimized example and can't reproduce it.
I will try again tomorrow.

However, I have my old logs from where it happened:

```
INFO: detected pid 1, running init handler
I0208 22:29:05.476855      10 main.go:322]  "level"=0 "msg"="starting up"  "args"=["/git-sync"]
I0208 22:29:06.392616      10 main.go:677]  "level"=0 "msg"="update required"  "local"="3f18518714b300c33fff38fd69a2ad64cfe51db4" "remote"="cc98f6d534f6d10c25db6655f0732727152f8d96" "rev"="HEAD"
I0208 22:29:06.392661      10 main.go:481]  "level"=0 "msg"="syncing git"  "hash"="cc98f6d534f6d10c25db6655f0732727152f8d96" "rev"="HEAD"
I0208 22:29:09.673844      10 main.go:502]  "level"=0 "msg"="adding worktree"  "branch"="origin/master" "path"="/tmp/rev-cc98f6d534f6d10c25db6655f0732727152f8d96"
E0208 22:29:09.673940      10 main.go:348]  "msg"="failed to sync repo, aborting" "error"="error running command: exit status 128: \"fatal: Invalid path '/gitrepo': No such file or directory\\nPreparing worktree (detached HEAD cc98f6d5)\\n\"" 
```
As you can see the path is "path"="/tmp/rev-... but then the error happens and /gitrepo is coming from somewhere.
Strange...

**Comment by yuranos** [comment]
Just in case: I'm using `k8s.gcr.io/git-sync:v3.1.5` and creating a deployment(GKE).

**Comment by thockin** [comment]
Strange indeed.  I can't fathom where that would be coming from.  If this
was actually a persistent volume, maybe the old worktree reference still
existed in the repo and the actual dir didn't?  E.g.

```
cat .git/worktrees/foobar/gitdir
/tmp/foobar/.git
```

Probably one of the commands we run doesn't handle that well, and kaboom.
If you can repro, please let me know.  I won't have time to try to repro
myself for a while.

**Comment by thockin** [comment]
This indeed related to the on-disk state of worktrees.  The worktree metadata has an absolute path baked in.  Now, I don't know WHY it is barfing I don't know.  My workstation has git 2.31, and debian only seems to have 2.20.  Curiously, 2.31 DOES NOT fail this way.   I can not find a way to make 2.20 NOT fail.

What I *can* do is pull git from the backports tree.  I'll work on a PR.  The only other alternative is to manually edit the worktree metadata to use a relative path, which may be a good idea ALSO.
